libs/core/type_support/include/hpx/type_support/relocate_at.hpp

```diff
@@ -0,0 +1,151 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/concepts/concepts.hpp>
+#include <hpx/type_support/construct_at.hpp>
+#include <hpx/type_support/is_relocatable.hpp>
+#include <hpx/type_support/is_trivially_relocatable.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+#include <memory>
+#endif
+
+namespace hpx::detail {
+#if __cplusplus <= 201703L
+    // pre c++17 std::destroy_at can be used only on non-array types
+    template <typename T,
+        HPX_CONCEPT_REQUIRES_(std::is_destructible_v<T> && !std::is_array_v<T>)>
+#else
+    // c++17 std::destroy_at can be used on array types, destructing each element
+    template <typename T, HPX_CONCEPT_REQUIRES_(std::is_destructible_v<T>)>
+#endif
+    struct destroy_guard
+    {
+        T* t;
+        explicit destroy_guard(T* t)
+          : t(t)
+        {
+        }
+        ~destroy_guard()
+        {
+            std::destroy_at(t);
+        }
+    };
+}    // namespace hpx::detail
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+using std::relocate;
+using std::relocate_at;
+#else
+
+namespace hpx::experimental {
+
+    namespace detail {
+
+        /*
+        The condition to use memmove is:
+                hpx::is_trivially_relocatable_v<T> &&
+                !std::is_volatile_v<T>
+
+        The reason for the volatile check is that memmove is not allowed to
+        change the value of a volatile object.
+        */
+
+        template <typename T>
+        constexpr bool relocate_using_memmove =
+            hpx::is_trivially_relocatable_v<T> && !std::is_volatile_v<T>;
+
+        template <typename T,
+            std::enable_if_t<relocate_using_memmove<T>, int> = 0>
+        T* relocate_at_helper(T* src, T* dst) noexcept
+        {
+            void* dst_void = const_cast<void*>(static_cast<const void*>(dst));
+            void* src_void = const_cast<void*>(static_cast<const void*>(src));
+
+            std::memmove(dst_void, src_void, sizeof(T));
+
+            return std::launder(dst);
+        };
+
+        // this is move and destroy
+        template <typename T,
+            std::enable_if_t<!relocate_using_memmove<T>, int> = 0>
+        T* relocate_at_helper(T* src, T* dst) noexcept(
+            // has non-throwing move constructor
```

Comment:
FWIW, I don't think this comment is useful. :)

Reply:
Looks like I was playing captain obvious there